pkg/controller/master/backup/backup_target.go

```diff
@@ -79,32 +80,68 @@ func (h *TargetHandler) OnBackupTargetChange(key string, setting *harvesterv1.Se
 		return setting, err
 	}
 
-	// Since we remove S3 access key id and secret access key after S3 backup target has been verified,
-	// we stop the controller to reconcile it again.
-	if target.Type == settings.S3BackupType && (target.SecretAccessKey == "" || target.AccessKeyID == "") {
-		return nil, nil
-	}
+	logrus.Debugf("backup target change:%s:%s", target.Type, target.Endpoint)
 
-	if err = h.updateLonghornTarget(target); err != nil {
-		return nil, err
-	}
+	switch target.Type {
+	case settings.S3BackupType:
+		// Since S3 access key id and secret access key are stripped after S3 backup target has been verified
+		// in reUpdateBackupTargetSettingSecret
+		// stop the controller to reconcile it again.
+		if target.SecretAccessKey == "" && target.AccessKeyID == "" {
+			return nil, nil
+		}
+
+		if err = h.updateLonghornTarget(target); err != nil {
+			return nil, err
+		}
 
-	if target.Type == settings.S3BackupType {
 		if err = h.updateBackupTargetSecret(target); err != nil {
 			return nil, err
 		}
-	}
 
-	return h.updateBackupTargetSetting(setting.DeepCopy(), target, err)
+		return h.reUpdateBackupTargetSettingSecret(setting, target, err)
+
+	case settings.NFSBackupType:
+		if err = h.updateLonghornTarget(target); err != nil {
+			return nil, err
+		}
+
+		// delete the may existing previous secret of S3
+		if err = h.deleteBackupTargetSecret(target); err != nil {
+			return nil, err
+		}
+
+		return nil, nil
+
+	default:
+		// reset backup target to default, then delete related settings
+		if target.IsDefaultBackupTarget() {
+			// default backup target is not usable, longhorn may not support default (blank) backup target
+			// delete it
+			if err = h.deleteLonghornTarget(target); err != nil {
+				return nil, err
+			}
+
+			// delete the may existing previous secret of S3
+			if err = h.deleteBackupTargetSecret(target); err != nil {
+				return nil, err
+			}
+
+			return nil, nil
+		}
+
+		return nil, fmt.Errorf("Invalid backup target type:%s or parameter", target.Type)
+	}
 }
 
-func (h *TargetHandler) updateBackupTargetSetting(setting *harvesterv1.Setting, target *settings.BackupTarget, err error) (*harvesterv1.Setting, error) {
+func (h *TargetHandler) reUpdateBackupTargetSettingSecret(setting *harvesterv1.Setting, target *settings.BackupTarget, err error) (*harvesterv1.Setting, error) {
 	harvesterv1.SettingConfigured.SetError(setting, "", err)
```

Comment:
`harvesterv1.SettingConfigured.SetError(setting, "", err)  `
--> when calls happen, err can only be nil.    because all errors have already been returned, no chance to here.
who/where  set the "SettingConfigured" when real error happens?
@FrankYang0529 @guangbochen

Reply:
Yes, it looks like err can only be nil before my last update. For me, it's okay to remove err parameter here.

https://github.com/harvester/harvester/pull/1447/commits/2c1b9c2c062dbad5e59df48af675182da1db7a7c#diff-b8a98fcbcd22961df4c633ac883a54f7e28155bcbecccd9a2aa966f004d07035